Clarify Ephemeral Containers RBAC (#55447)
kubernetes/website: content/en/docs/tasks/debug/debug-application/debug-running-pod.md

utilities, such as with [distroless images](
https://github.com/GoogleContainerTools/distroless).

### Permissions for ephemeral containers {#ephemeral-container-permissions}

Adding an ephemeral container to a running Pod uses the Pod's
`ephemeralcontainers` subresource. At the API level this is a `PATCH` (or
`UPDATE`) request on `pods/ephemeralcontainers`, which is governed by its own

> RBAC verbs and is *not* granted by `patch` on `pods`.

Reviewer: How about this:
```suggestion
For security reasons, adding an ephemeral container to a running Pod uses the Pod's
`ephemeralcontainers` subresource. At the API level this is an
**update** request on `pods/ephemeralcontainers`. Just being authorized to **patch**
a Pod does not allow users to add ephemeral containers to that Pod.
```

The `pods/ephemeralcontainers` subresource is intentionally not part of the
default `admin` and `edit`
[user-facing ClusterRoles](/docs/reference/access-authn-authz/rbac/#user-facing-roles).
For background on why, see
[Ephemeral Containers: Access control](/docs/concepts/workloads/pods/ephemeral-containers/#access-control).

If you run `kubectl debug` to add an ephemeral container without the required
permission, you will see an error similar to:

```
Error from server (Forbidden): pods "example" is forbidden: User "alice" cannot
patch resource "pods/ephemeralcontainers" in API group "" in the namespace "example-ns"
```

> Cluster administrators can grant this permission in one of two ways.

Reviewer: ```suggestion
If you're managing cluster access and your cluster uses [Kubernetes RBAC](/docs/reference/access-authn-authz/rbac/), you can grant this permission in one of two ways.
```

ClusterRole itself aggregates into `admin`:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: aggregate-ephemeralcontainers-edit
  labels:
    rbac.authorization.k8s.io/aggregate-to-edit: "true"
rules:
- apiGroups: [""]
  resources: ["pods/ephemeralcontainers"]
  verbs: ["patch", "update"]
```

To grant the permission only to specific users or groups (for example, a
dedicated debugging or SRE group), create a narrower ClusterRole and bind it
directly:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: debug-ephemeralcontainers
rules:
- apiGroups: [""]
  resources: ["pods/ephemeralcontainers"]
  verbs: ["patch", "update"]
```

{{< note >}}
Ephemeral containers are evaluated by [Pod Security admission](/docs/concepts/security/pod-security-admission/)
against the same [Pod Security Standards](/docs/concepts/security/pod-security-standards/)
as regular containers. When granting permission to add ephemeral containers to
Pods governed by the `baseline` or `restricted` policy, choose a matching
[debugging profile](#debugging-profiles) such as `--profile=baseline` or

> `--profile=restricted` with `kubectl debug`. Any validating admission webhooks in your cluster that do not evaluate the ephemeralContainers field may also need updating to evaluate the `ephemeralContainers` field in a Pod's spec.

Reviewer: ```suggestion
`--profile=restricted` with `kubectl debug`.

Any admission webhooks or admission policies in your cluster, that do not evaluate
the `ephemeralContainers` field, may also need updating to evaluate the `ephemeralContainers`
field in a Pod's spec.
```

utilities, such as with [distroless images](
https://github.com/GoogleContainerTools/distroless).

### Permissions for ephemeral containers {#ephemeral-container-permissions}

Adding an ephemeral container to a running Pod uses the Pod's
`ephemeralcontainers` subresource. At the API level this is a `PATCH` (or
`UPDATE`) request on `pods/ephemeralcontainers`, which is governed by its own

> RBAC verbs and is **not** granted by `patch` on `pods`.

Reviewer 2: single PR has two ways `not` is emphasized. Intentional?

```suggestion
RBAC verbs and is *not* granted by `patch` on `pods`.
```

Reviewer 2: feel free to keep bold, but prefer to keep consistent

ClusterRole itself aggregates into `admin`:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: aggregate-ephemeralcontainers-edit
  labels:
    rbac.authorization.k8s.io/aggregate-to-edit: "true"
rules:
- apiGroups: [""]
  resources: ["pods/ephemeralcontainers"]
  verbs: ["patch", "update"]
```

To grant the permission only to specific users or groups (for example, a
dedicated debugging or SRE group), create a narrower ClusterRole and bind it
directly:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: debug-ephemeralcontainers
rules:
- apiGroups: [""]
  resources: ["pods/ephemeralcontainers"]
  verbs: ["patch", "update"]
```

{{< note >}}
Ephemeral containers are evaluated by [Pod Security admission](/docs/concepts/security/pod-security-admission/)
against the same [Pod Security Standards](/docs/concepts/security/pod-security-standards/)
as regular containers. When granting permission to add ephemeral containers to
Pods governed by the `baseline` or `restricted` policy, choose a matching
[debugging profile](#debugging-profiles) such as `--profile=baseline` or

> `--profile=restricted`. Any validating admission webhooks in your cluster that

Reviewer 2: ```suggestion
`--profile=restricted` with `kubectl debug`. Any validating admission webhooks in your cluster that
```

> were written before this API existed may also need to be updated to evaluate

Reviewer 2: Make it evergreen:

```suggestion
`--profile=restricted`. Any validating admission webhooks in your cluster that do not evaluate the ephemeralContainers field may also need updating to evaluate
```

    rbac.authorization.k8s.io/aggregate-to-edit: "true"
rules:
- apiGroups: [""]
  resources: ["pods/ephemeralcontainers"]
  verbs: ["patch", "update"]
```

To grant the permission only to specific users or groups (for example, a
dedicated debugging or SRE group), create a narrower ClusterRole and bind it
directly:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: debug-ephemeralcontainers
rules:
- apiGroups: [""]
  resources: ["pods/ephemeralcontainers"]
  verbs: ["patch", "update"]
```

{{< note >}}
Ephemeral containers are evaluated by [Pod Security admission](/docs/concepts/security/pod-security-admission/)
against the same [Pod Security Standards](/docs/concepts/security/pod-security-standards/)
as regular containers. When granting permission to add ephemeral containers to
Pods governed by the `baseline` or `restricted` policy, choose a matching
[debugging profile](#debugging-profiles) such as `--profile=baseline` or
`--profile=restricted` with `kubectl debug`.

Any admission webhooks or admission policies in your cluster, that do not evaluate
the `ephemeralContainers` field, may also need updating to evaluate the `ephemeralContainers`

> field in a Pod's spec.

Reviewer 3: ```suggestion
the `ephemeralContainers` field, may also need updating to intercept the
`pods/ephemeralcontainers` resource and evaluate the `ephemeralContainers` field in a Pod's spec.
```
